TripalCultivate/TripalCultivate-Phenotypes
# Add cases to valid file delimiter validator

**Issue #115 ** G4.115 Case strings in ValidDelimitedFile validator may need reconsidering

## Motivation
Create a more specific case the valid file validator to report if line has insufficient number of columns or exceeds number of columns expected.



## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an automated test to ensure it doesn't return.*

1. Create a case for lines values that exceed expected number of columns.
2. Create a case for line values that is less than the expected number of columns.
3. Add test scenarios to test new cases.

## Automated test
Class: DataFileDelimiterTest - testDataFileRowIsDelimited()

1. Raw row is an empty string.
2. None of the supported delimiters was used.
3. Unexpected number of columns (more columns returned with strict comparison set to true).
4. Unexpected number of columns (less columns returned with strict comparison set to true).
5. Unexpected number of columns (less columns returned with strict comparison set to false).
6. Line has 2 delimiters detected
7. Valid line expecting 4 values and strict comparison.
8. Valid line expecting at least 3 values and not strict comparison.
9. Line has one value and not strict comparison.
10. Line has one value and strict comparison.

## Manual test
```
// Setup
$validator = \Drupal::service('plugin.manager.trpcultivate_validator')
  ->createInstance('valid_delimited_file');

$validator->setSupportedMimeTypes([
  'tsv',
  'txt',
]);

$validator->setFileMimeType('text/tab-separated-values');

// Comparison Flag:

// Adjust values:
// Params: number of columns expected and comparison flag.
$number_of_columns = 4;
$comparison_flag = TRUE;


$validator->setExpectedColumns($number_of_columns, $comparison_flag);

// Test
$data_values = [
  'Value 1',
  'Value 2',
  'Value 3',
  'Value 4'
];

$raw_row_input = implode("\t", $data_values);
$validation_status = $validator->validateRawRow($raw_row_input);
dpm($validation_status);
```

Paste this code into a fully setup Tripal site with phenotypes modules installed and activated.

Comparison Flag:
If set to true, the delimiter validator must match the exact number of columns.
If set to false, the delimiter validator must match at least the number of columns.
